tensornetwork/network_components.py

```diff
@@ -1171,7 +1173,12 @@ def _flatten_trace_edges(edges: List[Edge],
   edge2 = Edge(node1=node, axis1=len(perm_front) + 1, name="TraceBack")
   node.edges = node.edges[:len(perm_front)] + [edge1, edge2]
   new_edge = connect(edge1, edge2, new_edge_name)
-  node.axis_names = [str(n) for n in range(len(node.edges))]
+  if has_nongeneric_names:
+    node.axis_names = (node.axis_names[0:len(perm_front)] + 
+                       [node.axis_names[new_edge.axis1], 
+                        node.axis_names[new_edge.axis2]])
+  else:
+    node.axis_names = [str(n) for n in range(len(node.edges))]
```

Comment:
Here the axis names will be preserved if they have non-generic names. Grab non flattened axis names and then the axes of the first edges amongst the edges being flattened.

If the names are generic a new set of names will be generated instead of just using the generic names that have been scrambled by _flatten_trace_edges. But I can remove this if you think its better to not regenerate the generic names.